File: DailyChallenges/Scala/LeetCode_1339.scala
object Solution {
    def maxProduct(root: TreeNode): Int = {
        def sum(n: TreeNode): Long = if (n == null) 0L else n.value + sum(n.left) + sum(n.right)
        val total = sum(root)
        
        var ans = 0L
        def search(n: TreeNode): Long = if (n == null) 0L else {
            val s = n.value + search(n.left) + search(n.right)
            ans = ans max (s * (total - s))
            s
        }
        
        search(root)
        (ans % (1e9.toInt + 7)).toInt
    }
}

File: DailyChallenges/CPP/LeetCode_1339.cpp
// 1339. Maximum Product of Splitted Binary Tree @ 2022/12/10
class Solution {
public:
    int maxProduct(TreeNode* root) {
        vector<long long> sums;
        function<long long(TreeNode*)> dfs = [&](TreeNode* node){
            long long sum = node->val;
            if (node->left) sum += dfs(node->left);
            if (node->right) sum += dfs(node->right);
            sums.push_back(sum);
            return sum;
        };
        long long total = dfs(root);
        sort(sums.begin(), sums.end());
        auto itr = lower_bound(sums.begin(), sums.end(), total/2);
        long long ans = *itr * (total - *itr);
        if (itr-- != sums.begin())
            ans = max(ans, *itr * (total - *itr));
        return ans % 1000000007;
    }
};
